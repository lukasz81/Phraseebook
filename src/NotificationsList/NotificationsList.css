figure {
  margin: 0 10px 0 0;
  padding: 0; }

.notifications-list ul {
  list-style-type: none;
  margin: 0;
  padding: 0; }

.notifications-list ul li {
  box-shadow: 0 1px 0 0 #ddd;
  padding: 10px 0;
  transition: all 0.2s ease-in-out; }

.notifications-list ul li.notification-container {
  cursor: pointer; }

.notifications-list ul li.notification-container:hover {
  color: #333; }

.notifications-list ul li.notification-container .text-container {
  position: relative;
  display: block;
  color: #838485;
  height: calc(2em + 5px);
  overflow: hidden;
  white-space: normal; }

.notifications-list ul li.notification-container .text-content {
  word-break: break-all;
  position: relative;
  display: block;
  max-height: 3em; }

.notifications-list ul li.notification-container .ellipsis {
  position: absolute;
  right: 1px;
  top: calc(4em + 2px - 100%);
  text-align: left;
  padding-left: 35px;
  background: linear-gradient(to right, #ffffff00, white 65%); }
