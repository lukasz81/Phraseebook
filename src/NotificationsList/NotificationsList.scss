@import "../variables.scss";

figure {
  margin: 0 10px 0 0;
  padding: 0;
}
.notifications-list {
  ul {
    list-style-type: none;
    margin:0;
    padding:0;
    li {
      box-shadow: 0 1px 0 0 #ddd;
      padding: 10px 0;
      transition: all 0.2s ease-in-out;
      &.notification-container {
        cursor: pointer;
        &:hover {
          color: $main-color-hover
        }
        .text-container {
          position: relative;
          display: block;
          color: #838485;
          height: calc(2em + 5px);
          overflow: hidden;
          white-space: normal;
        }
        .text-content {
          word-break: break-all;
          position: relative;
          display: block;
          max-height: 3em;
        }
        .ellipsis {
          position: absolute;
          right: 1px;
          top: calc(4em + 2px - 100%);
          text-align: left;
          padding-left: 35px;
          background: linear-gradient(to right, #ffffff00, white 65%);
        }
        // text-container gets a fixed height which acts as a viewport. Give it a few extra pixels since characters that extend below the word-line will appear cut off. (2em in this example with 5px extra)
        // text-content needs a maximum height which should be 1 line height greater than the viewport (3em in this example)
        // ellipsis needs to move inversely with the height of the text-node. So it is offset out of bounds of the viewport initially and then as the content grows, it moves up and will stop moving up when text-content hits its max-height. (Top starts at 1 height below the viewport bottom. Final position is at 1em.)
        // A CODE PEN BY Michael Chen
      }
    }
  }
}