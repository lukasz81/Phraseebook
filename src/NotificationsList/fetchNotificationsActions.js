import * as actionTypes from '../Redux/actionTypes';
const URL = 'http://www.mocky.io/v2/5b4315f12e00004c002230c3';
export const fetchTrackersAction = (dispatch) => {

    fetch(URL)
        .then(
            response => response.json().then(
                data => {
                    return dispatch({
                                type: actionTypes.FETCH_SUCCESS,
                                isWaiting: false,
                                notifications: data
                            })
                }
            )
        )
        .catch(
            error => {
                return dispatch({
                    type: actionTypes.FETCH_FAIL,
                    isWaiting: false,
                    error: error.message
                })}
        );

    return dispatch({
        type: actionTypes.FETCH_REQUEST,
        isWaiting: true
    })
};