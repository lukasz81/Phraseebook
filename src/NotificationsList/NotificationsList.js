import React, {Component} from 'react';
import {connect} from 'react-redux';
import {fetchNotificationsAction, markAsReadActions} from './notificationsActions';
import './NotificationsList.css';

export class NotificationsList extends Component {

    componentWillMount() {
        // Load content if not in store.
        if (!this.props.notifications) this.props.fetchNotificationsAction();
        if (this.props.notifications) this.props.markAsRead();
    }

    notificationClickHandler(event, id) {
        event.preventDefault();
        this.props.history.push('/notification/' + id)
    }

    createCopyForNotifications(notification) {
        const notificationTypeIsLike = notification.type === 'Like';
        const source = notificationTypeIsLike ? notification.likes : notification.comments;
        const maxNames = 2;
        const names = source.map(
            (item, index) => index < maxNames ? ` ${item.name}` : null
        );
        const trimmedBase = `${names}`.replace(/([,\s]+$)/g, '');
        const action = notificationTypeIsLike ? 'liked' : 'commented on';
        const multipleActionCopy =
            <span className={'two-line-tex'}>
                <b>{trimmedBase} and {source.length - maxNames} others</b> {action} your post: "{notification.post.title}."
            </span>;
        const singleActionCopy =
            <span className={'two-line-tex'}>
                <b>{names.length && !notificationTypeIsLike ? names : 'User'}</b> {action} your post: "{notification.post.title}."
            </span>;
        return source.length > 1 ? multipleActionCopy : singleActionCopy;

    }

    createListOfNotifications(notifications) {
        return notifications.map((notification, index) => {
                return (
                    <li
                        onClick={event => this.notificationClickHandler(event, notification.post.id)}
                        className={'flex-container notification-container'}
                        key={index}>
                        <figure style={{width: '50px'}}>
                            <img style={{width: '50px'}} className={'profile-picture'} src={'/img-placement.jpg'}
                                 alt={'Profile picture for ' + notification.type}/>
                        </figure>
                        {this.createCopyForNotifications(notification)}
                    </li>
                )
            }
        )
    }

    render() {
        const notifications = this.props.notifications;
        return (
            <div className={'default-padding'}>
                {notifications ?
                    <ul>
                        {this.createListOfNotifications(notifications)}
                    </ul>
                    :
                    <p>Loading notifications ...</p>
                }
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchNotificationsAction: () => {
            fetchNotificationsAction(dispatch)
        },
        markAsRead: () => {
            markAsReadActions(dispatch)
        }
    }
};

const mapStateToProps = (state) => {
    return {
        isWaiting: state.isWaiting,
        notifications: state.notifications
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(NotificationsList);