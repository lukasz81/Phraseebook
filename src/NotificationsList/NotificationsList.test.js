import NotificationsList from './notificationsList';
import Enzyme, {shallow} from "enzyme";
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({adapter: new Adapter()});

describe('Tests notifications list component', () => {

    const Comment = {
        type: 'Comment',
        post: {
            id: 'b1638f970c3ddd528671df76c4dcf13e',
            title: 'Acme Inc dynamically scales niches worldwide'
        },
        comments: [
            {
                id: '5497afbf9df3f6ff6f9ba11cdef5310f',
                name: 'Suoma Narjus',
                commentText: 'Acme remains one of my fave company ever! The way they scale is so dynamic that makes HTML5 look static!'
            }
        ],
        isRead: true
    };

    const Like = {
        type: 'Like',
        post: {
            id: 'b1638f970c3ddd528671df76c4dcf13e',
            title: 'Acme Inc dynamically scales niches worldwide'
        },
        likes: [
            {
                id: '403f220c3d413fe9cb0b36142ebfb35d',
                name: 'Mary T. Price'
            },
            {
                id: '5497afbf9df3f6ff6f9ba11cdef5310f',
                name: 'Suoma Narjus'
            },
            {
                id: 'fa527981cbbcb070be95854985c3188f',
                name: 'Sandra Ortega'
            },
            {
                id: '4c18d43d4deccbac21a26c55f1033f53',
                name: 'William Hunt'
            },
            {
                id: '38be3079117301f2f61264d6e0fbf7db',
                name: 'An Mao'
            }
        ],
        isRead: true
    };

    const mockFuncFetchNotifications = jest.fn();
    const mockFuncMarkAsRead = jest.fn();
    const pushMock = jest.fn();
    const historyMock = { push: pushMock };

    const makeComponent = (likeObject = Like, commentObject = Comment) => {
        return shallow(
            <NotificationsList.WrappedComponent
                history={historyMock}
                notifications={[likeObject,commentObject]}
                fetchNotificationsAction={mockFuncFetchNotifications}
                markAsRead={mockFuncMarkAsRead}
            />
        )
    };

    it('should render correctly', () => {
        expect(makeComponent()).toMatchSnapshot();
    });

    it('should call componentWillMount and call through dependent functions', () => {
        expect(mockFuncFetchNotifications).toHaveBeenCalledTimes(0);
        expect(mockFuncMarkAsRead).toHaveBeenCalledTimes(1);
    });

    it('should call prevent default when calling notificationClickHandler', () => {
        const mockPrevent = jest.fn();
        const event = {
            preventDefault: mockPrevent
        };
        const component = makeComponent();
        component.instance().notificationClickHandler(event,'123');
        expect(mockPrevent).toHaveBeenCalledTimes(1);
    });

    it('should call history push when calling notificationClickHandler', () => {
        const event = {
            preventDefault: jest.fn()
        };
        const component = makeComponent();
        component.instance().notificationClickHandler(event,'123');
        expect(pushMock).toHaveBeenCalled();
        expect(historyMock.push.mock.calls[0]).toEqual(['/notification/123']);
    });

    it('should create list of notifications', () => {
        const component = makeComponent();
        component.instance().createListOfNotifications([Like,Comment]);
        expect(component.find('li').length > 0).toBe(true)
    });



});