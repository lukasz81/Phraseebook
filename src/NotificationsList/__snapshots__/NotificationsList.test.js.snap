// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests notifications list component should render correctly 1`] = `
<div
  className="default-padding notifications-list"
>
  <ul>
    <li
      className="flex-container notification-container"
      key="0"
      onClick={[Function]}
    >
      <figure
        style={
          Object {
            "width": "50px",
          }
        }
      >
        <img
          alt="Profile picture for Like"
          className="profile-picture"
          src="/img-placement.jpg"
          style={
            Object {
              "width": "50px",
            }
          }
        />
      </figure>
      <div
        className="text-container"
      >
        <span
          className="two-line-tex text-content"
        >
          <b>
             Mary T. Price, Suoma Narjus
             and 
            3
             others
          </b>
           
          liked
           your post: "
          Acme Inc dynamically scales niches worldwide
          ."
          <div
            className="ellipsis"
          >
            ...
          </div>
        </span>
      </div>
    </li>
    <li
      className="flex-container notification-container"
      key="1"
      onClick={[Function]}
    >
      <figure
        style={
          Object {
            "width": "50px",
          }
        }
      >
        <img
          alt="Profile picture for Comment"
          className="profile-picture"
          src="/img-placement.jpg"
          style={
            Object {
              "width": "50px",
            }
          }
        />
      </figure>
      <div
        className="text-container"
      >
        <span
          className="two-line-tex text-content"
        >
          <b>
             Suoma Narjus
          </b>
           
          commented on
           your post: "
          Acme Inc dynamically scales niches worldwide
          ."
          <div
            className="ellipsis"
          >
            ...
          </div>
        </span>
      </div>
    </li>
  </ul>
</div>
`;
